In [40]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import AdaBoostClassifier
import warnings
warnings.filterwarnings("ignore")  # Suppress all warnings

In [9]:
df = pd.read_csv(r'C:\Users\Admin\Desktop\data science assighnments\Random Forest\Random Forest\random forest\data.csv')

In [10]:
df.head()

Id number       RI     Na    Mg    Al     Si     K    Ca   Ba   Fe  \
0          1  1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.0  0.0   
1          2  1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.0  0.0   
2          3  1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.0  0.0   
3          4  1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.0  0.0   
4          5  1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.0  0.0   

   Type of glass  
0              1  
1              1  
2              1  
3              1  
4              1

In [11]:
df.describe()

Id number          RI          Na          Mg          Al          Si  \
count  214.000000  214.000000  214.000000  214.000000  214.000000  214.000000   
mean   107.500000    1.518365   13.407850    2.684533    1.444907   72.650935   
std     61.920648    0.003037    0.816604    1.442408    0.499270    0.774546   
min      1.000000    1.511150   10.730000    0.000000    0.290000   69.810000   
25%     54.250000    1.516522   12.907500    2.115000    1.190000   72.280000   
50%    107.500000    1.517680   13.300000    3.480000    1.360000   72.790000   
75%    160.750000    1.519157   13.825000    3.600000    1.630000   73.087500   
max    214.000000    1.533930   17.380000    4.490000    3.500000   75.410000   

                K          Ca          Ba          Fe  Type of glass  
count  214.000000  214.000000  214.000000  214.000000     214.000000  
mean     0.497056    8.956963    0.175047    0.057009       2.780374  
std      0.652192    1.423153    0.497219    0.097439       2.103739  
min      0.000000    5.430000    0.000000    0.000000       1.000000  
25%      0.122500    8.240000    0.000000    0.000000       1.000000  
50%      0.555000    8.600000    0.000000    0.000000       2.000000  
75%      0.610000    9.172500    0.000000    0.100000       3.000000  
max      6.210000   16.190000    3.150000    0.510000       7.000000

In [12]:
df.isnull().sum()

Id number        0
RI               0
Na               0
Mg               0
Al               0
Si               0
K                0
Ca               0
Ba               0
Fe               0
Type of glass    0
dtype: int64

In [15]:

# Drop the 'Id' column as it's not needed for classification
df.drop("Id number", axis=1, inplace=True)


In [18]:

# Split the features and target variable
X = df.iloc[:, :-1]  # Features
y = df["Type of glass"]       # Target variable


In [21]:

# Standardize the feature data
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

In [23]:

# Split into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.3, random_state=42, stratify=y)


In [24]:
# Train the Random Forest Classifier
rf_classifier = RandomForestClassifier(n_estimators=100, random_state=42)
rf_classifier.fit(X_train, y_train)


RandomForestClassifier(random_state=42)

In [25]:
# Make predictions
y_pred = rf_classifier.predict(X_test)


In [26]:

# Evaluate the model
print("Classification Report:")
print(classification_report(y_test, y_pred))

print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))

print("\nAccuracy Score:")
print(accuracy_score(y_test, y_pred))

Classification Report:
              precision    recall  f1-score   support

           1       0.80      0.95      0.87        21
           2       0.86      0.78      0.82        23
           3       1.00      0.40      0.57         5
           5       0.60      0.75      0.67         4
           6       0.75      1.00      0.86         3
           7       1.00      0.89      0.94         9

    accuracy                           0.83        65
   macro avg       0.83      0.80      0.79        65
weighted avg       0.85      0.83      0.83        65


Confusion Matrix:
[[20  1  0  0  0  0]
 [ 2 18  0  2  1  0]
 [ 3  0  2  0  0  0]
 [ 0  1  0  3  0  0]
 [ 0  0  0  0  3  0]
 [ 0  1  0  0  0  8]]

Accuracy Score:
0.8307692307692308


In [34]:
# Apply BaggingClassifier
bagging = BaggingClassifier(
    estimator=DecisionTreeClassifier(), 
    n_estimators=50, 
    random_state=42
)
bagging.fit(X_train, y_train)



BaggingClassifier(estimator=DecisionTreeClassifier(), n_estimators=50,
                  random_state=42)

In [35]:

# Predict and evaluate
y_pred = bagging.predict(X_test)
print("Bagging Classifier Accuracy:", accuracy_score(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred))

Bagging Classifier Accuracy: 0.8307692307692308

Classification Report:
               precision    recall  f1-score   support

           1       0.79      0.90      0.84        21
           2       0.90      0.78      0.84        23
           3       0.67      0.40      0.50         5
           5       0.60      0.75      0.67         4
           6       1.00      1.00      1.00         3
           7       0.90      1.00      0.95         9

    accuracy                           0.83        65
   macro avg       0.81      0.81      0.80        65
weighted avg       0.83      0.83      0.83        65



In [41]:
# Apply AdaBoostClassifier
adaboost = AdaBoostClassifier(
    estimator=DecisionTreeClassifier(max_depth=1), 
    n_estimators=50, 
    learning_rate=1.0, 
    random_state=42
)
adaboost.fit(X_train, y_train)


AdaBoostClassifier(estimator=DecisionTreeClassifier(max_depth=1),
                   random_state=42)

In [42]:
# Predict and evaluate
y_pred_boost = adaboost.predict(X_test)
print("AdaBoost Classifier Accuracy:", accuracy_score(y_test, y_pred_boost))
print("\nClassification Report:\n", classification_report(y_test, y_pred_boost))

AdaBoost Classifier Accuracy: 0.24615384615384617

Classification Report:
               precision    recall  f1-score   support

           1       0.00      0.00      0.00        21
           2       0.30      0.26      0.28        23
           3       0.12      0.80      0.21         5
           5       0.00      0.00      0.00         4
           6       0.60      1.00      0.75         3
           7       0.50      0.33      0.40         9

    accuracy                           0.25        65
   macro avg       0.25      0.40      0.27        65
weighted avg       0.21      0.25      0.20        65



findings :- Bagging-
Accuracy: 83.1%. Strong performance across most classes, especially for classes 1, 2, 6, and 7.Class 6 performs perfectly with 100% precision and recall.
Boosting:-
Accuracy: 24.6%
Poor performance for most classes, especially for classes 1, 5, and 7.
Class 3 has high recall (80%), but it doesn't translate into good overall performance due to poor precision and f1-score.

Question Answers:-

1. Bagging creates multiple models (learners) by training each model independently on a randomly sampled subset of the training data, with replacement (bootstrap sampling).
After training, the predictions of these models are combined (typically by averaging for regression or majority voting for classification) to make a final prediction.

Boosting trains multiple models sequentially, with each new model focusing on correcting the mistakes made by the previous model.
Each model is trained with a re-weighted dataset, where incorrectly predicted samples from the previous model are given higher weights.

2. we can overcome imbalance data with help of resampling techniques. like  Oversampling the Minority Class, Undersampling the Majority Class, Combination of Over-sampling and Under-sampling (Hybrid). or with the help of ensemble methods or evaluation metrices. 

